### PR TITLE
Schema annotate nested sequence and dict

### PR DESCRIPTION
**Pull request by blackboxsw**
## rebase merge, not a squash merge

Fix schema annotation handling and schema marks for nested sequences and maps.

```
    schema: handle nested dicts and lists in schema marks and annotation
    
    YAML schema processing of sequences will process the root node and
    then leaf-first processing of children. This adds complexity for nested
    maps and sequences that may live under intermediate parents below the
    root because those parents have not yet been processed.
    
    Refactor safeyaml schemamark handling to cope with depth-first
    processing of nested mappings and sequences.
    
    Make SchemaPathMarks a class instead of namedtuple to allow for
    operator overloading of equality and contains operators.
    
    Add helper private functions to correct the parentage of
    schema marks during construct_sequence and construct_mapping calls
    when the newly schema mark happens to be a parent of children
    stored in schemamarks_by_line.

---
schema: support cloud-init schema in early boot or in dev environ

Add DataSouceNotFoundException handling in cloudinit schema
subcommand which prints a warning when no datsource cache is
present.

Allows developers to run the schema subcommand directly from
the repository with:
 PYTHONPATH=. python3 -m cloudinit.cmd.main schema -c <your_yaml>
```

## Test Steps
```
cat > invalid_nested.yaml <<EOF
#cloud-config
packages:
- apta: [bogus]
EOF

PYTHONPATH=. python3 -m cloudinit.cmd.main schema --annotate -c invalid_nested.yaml
# 1. expect trace back on datasource not found when on main
# 2. cherry-pick the first commit from this PR, to allow for development run of cloud-init schema --annotate
#      - expect Traceback on KeyError: 'packages.0' becuase schema 
# 3. with both commits from this PR:
$ PYTHONPATH=. python3 -m cloudinit.cmd.main schema --annotate -c invalid_nested.yaml
#cloud-config
packages:
- apta: [bogus]		# E1,E2

# Errors: -------------
# E1: {'apta': ['bogus']} is not of type 'array'
# E2: {'apta': ['bogus']} is not valid under any of the given schemas

```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/safeyaml.py [hunk 1]
```diff
@@ -15,34 +15,101 @@
...
+        ):
+            return False
+        return True
+
+    def __eq__(self, other):
+        return (
+            self.start_mark.line == other.start_mark.line
+            and self.start_mark.column == other.start_mark.column
+            and self.end_mark.line == other.end_mark.line
+            and self.start_mark.line == other.start_mark.line
```
Comment (TheRealFalcon) [inline cloudinit/safeyaml.py:48]: These last 3 lines are all duplicate conditions, correct?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Really nice change here!

I left a couple minor inline comments.

File: cloudinit/cmd/devel/__init__.py [hunk 2]
```diff
@@ -27,7 +28,10 @@ def read_cfg_paths(fetch_existing_datasource: str = "") -> Paths:
     """
     init = Init(ds_deps=[])
     if fetch_existing_datasource:
-        init.fetch(existing=fetch_existing_datasource)
+        try:
+            init.fetch(existing=fetch_existing_datasource)
+        except DataSourceNotFoundException:
+            pass
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/cmd/devel/__init__.py:34]: Good pt, pushed this try/except into schema.py and print a WARNING on the commandline when if no cache is present to inform user about the situation when running cloud-init schema without a datasource available.

File: cloudinit/safeyaml.py (hunk 1 above)
Comment (blackboxsw) [reply cloudinit/safeyaml.py]: ahh you are right, I've dropped two and changed the last one to `self.end_mark.column == other.end_mark.column`

**Review by TheRealFalcon — APPROVED** [review]
LGTM! I'll leave it to you to merge.

**Pull request merged**